sumpy/tools.py

```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
                 if (np.array(mi) >= np.array(other_mi)).all()),
             key=lambda other_mi: sum(self.mi_dist(mi, other_mi)))
 
+
+class LaplaceDerivativeTaker(MiDerivativeTaker):
+
+    def __init__(self, expr, var_list, rscale=1, sac=None):
+        super(LaplaceDerivativeTaker, self).__init__(expr, var_list, rscale, sac)
+        self.scaled_var_list = [add_to_sac(self.sac, v/rscale) for v in var_list]
+        self.scaled_r = add_to_sac(self.sac,
+                sym.sqrt(sum(v**2 for v in self.scaled_var_list)))
+
+    def diff(self, mi):
+        # Return zero for negative values. Makes the algorithm readable.
+        if min(mi) < 0:
+            return 0
+        try:
+            return self.cache_by_mi[mi]
+        except KeyError:
+            pass
+
+        dim = self.dim
+        if max(mi) == 1:
+            return MiDerivativeTaker.diff(self, mi)
+        d = -1
+        for i in range(dim):
+            if mi[i] >= 2:
+                d = i
+                break
+        assert d >= 0
+        expr = 0
+        for i in range(dim):
+            mi_minus_one = list(mi)
+            mi_minus_one[i] -= 1
+            mi_minus_one = tuple(mi_minus_one)
+            mi_minus_two = list(mi)
+            mi_minus_two[i] -= 2
+            mi_minus_two = tuple(mi_minus_two)
+            x = self.scaled_var_list[i]
+            n = mi[i]
+            if i == d:
+                if dim == 3:
+                    expr -= (2*n - 1) * x * self.diff(mi_minus_one)
+                    expr -= (n - 1)**2 * self.diff(mi_minus_two)
+                else:
+                    expr -= 2 * x * (n - 1) * self.diff(mi_minus_one)
+                    expr -= (n - 1) * (n - 2) * self.diff(mi_minus_two)
+                    if n == 2 and sum(mi) == 2:
+                        expr += 1
+            else:
+                expr -= 2 * n * x * self.diff(mi_minus_one)
+                expr -= n * (n - 1) * self.diff(mi_minus_two)
+        expr /= self.scaled_r**2
+        expr = add_to_sac(self.sac, expr)
+        self.cache_by_mi[mi] = expr
+        return expr
+
+
+class RadialDerivativeTaker(MiDerivativeTaker):
+
+    def __init__(self, expr, var_list, rscale=1, sac=None):
+        """
+        Takes the derivatives of a radial function.
+        """
+        import sumpy.symbolic as sym
+        super(RadialDerivativeTaker, self).__init__(expr, var_list, rscale, sac)
+        empty_mi = (0,) * len(var_list)
+        self.cache_by_mi_q = {(empty_mi, 0): expr}
+        self.r = sym.sqrt(sum(v**2 for v in var_list))
+        rsym = sym.Symbol("_r")
+        r_expr = expr.xreplace({self.r**2: rsym**2})
+        self.is_radial = not any(r_expr.has(v) for v in var_list)
+        self.var_list_multiplied = [add_to_sac(sac, v * rscale) for v in var_list]
+
+    def diff(self, mi, q=0):
+        """
+        See [1] for the algorithm
```

Comment:
Talk a little bit about the properties and why this was chosen here.

Reply:
Done in 113c9643507884756aa3dfa8782b46fbb87652e7